SolarTracker/Pages/Index.razor

```diff
@@ -0,0 +1,362 @@
+        string? timeZoneId = arg.Value?.ToString();
+        if (timeZoneId is null) return;
+        _timeZoneInfo = _timeZoneInfos.FirstOrDefault(tzi => tzi.Id == timeZoneId);
+        if (_timeZoneInfo is null) return;
+        _timeZoneOffset = _timeZoneInfo.GetUtcOffset(_selectedDateTime);
+        _selectedDateTime = _selectedDateTime.Add(_timeZoneOffset);
+    }
+
+    private async Task OnExtentChanged()
+    {
+        await GetCenterAndTimeZone();
+    }
+
+    private async Task GoToSunrise()
+    {
+        _movingToSunrise = true;
+        _selectedDateTime = _selectedDateTime.AddMinutes(_skipInterval);
+        await OnMapRendered();
+    }
+
+    private async Task GoToSunset()
+    {
+        _movingToSunset = true;
+        _selectedDateTime = _selectedDateTime.AddMinutes(_skipInterval);
+        await OnMapRendered();
+    }
+
+    private async Task MoveTimeUntilStateChanged(int minutesInterval, bool forward, bool isIntersected)
+    {
+        _selectedDateTime = _selectedDateTime.AddMinutes(forward ? minutesInterval : -minutesInterval);
+        await CreateTerminator();
+        bool nextDirection = forward;
+        bool nextIntersected = isIntersected;
+        if (isIntersected != await GeometryEngine.Intersects(_center!, _polygon!))
+        {
+            nextDirection = !forward;
+            nextIntersected = !isIntersected;
+        }
+
+        if (minutesInterval > 1)
+        {
+            await MoveTimeUntilStateChanged((int)(minutesInterval / 2.0), nextDirection, nextIntersected);
+        }
+    }
+
+    private async Task FetchTimeZoneData()
+    {
+        try
+        {
+            var client = new HttpClient();
+            HttpResponseMessage result;
+            
+            List<Task> fetchTasks = new();
+            for (int i = 1; i < 40; i++)
+            {
+                int featureId = i;
+                fetchTasks.Add(Task.Run(async () =>
+                {
+                    try
+                    {
+                        result = await client.GetAsync($"{_timeZoneServiceUrl}{featureId}?f=json");
+                        if (result.IsSuccessStatusCode)
+                        {
+                            TimeZoneData data = (await result.Content.ReadFromJsonAsync<TimeZoneData>())!;
+                            _timeZoneData.Add(data.Feature);
+                            featureId++;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine(ex);
+                    }
+                }));
+            }
+            
+            await Task.WhenAll(fetchTasks);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex.ToString());
+        }
+        _timeZonesLoaded = true;
+        await GetCenterAndTimeZone();
+    }
+
+    private async Task GetCenterAndTimeZone()
+    {
+        _center = await _mapView!.GetCenter();
```

Comment:
I'm not sure what you're suggesting here. Calling this from `OnExtentChanged` is causing the lag, yes. My thought was to tweak the timeout on how often that fires. What would a layer query look like here?

Reply:
Right now you are downloading the geometries and then looping through them, and testing the intersect on each one. I was suggesting that it might be faster/more efficient to simply ask the service to do all of that work for you and return only the timezone that the point is currently in. If you look at this [query](https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/World_Time_Zones/FeatureServer/0//query?where=1%3D1&objectIds=&time=&geometry=-122.614408%2C45.5377835&geometryType=esriGeometryPoint&inSR=4326&spatialRel=esriSpatialRelIntersects&resultType=none&distance=0.0&units=esriSRUnit_Meter&relationParam=&returnGeodetic=false&outFields=*&returnGeometry=true&returnCentroid=false&featureEncoding=esriDefault&multipatchOption=xyFootprint&maxAllowableOffset=&geometryPrecision=&outSR=&defaultSR=&datumTransformation=&applyVCSProjection=false&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnExtentOnly=false&returnQueryGeometry=false&returnDistinctValues=false&cacheHint=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&having=&resultOffset=&resultRecordCount=&returnZ=false&returnM=false&returnExceededLimitFeatures=true&quantizationParameters=&sqlFormat=none&f=html&token=) it asks for just the one timezone at that location.

My idea was that on pan you just fire off that one request and get the result back. But you are right, if the individual performance of looping/intersecting is fine, then just putting in the delay probably takes care of it. And it does seem much faster now! I am curious though how much time this method takes to run though, we could compare it to just asking the service.

Reply:
oh, another thing you could do here is to update the center point on every pan, that way the black dot was always centered as the user pans, but then delay the actual computation until they stop panning, or using the delay technique you are using already.